packages/grid/x-data-grid-pro/src/useDataGridProComponent.tsx

```diff
@@ -56,6 +59,7 @@ export const useDataGridProComponent = (apiRef: GridApiRef, props: DataGridProPr
   useGridPrintExport(apiRef, props);
   useGridClipboard(apiRef);
   useGridDimensions(apiRef, props);
+  useGridDetailPanel(apiRef, props);
```

Comment:
We should be able to put this hook next to `useGridTreeData`, before `useGridColumns`
That way, the 1st  sync state init in `useGridColumns` already have the master detail panel pre-processor registered.
That way you have 2 pre-processing on mount instead of 3.
The more we move forward with pre-processing, the more I think `useGridColumns` and `useGridRows` should be late in the call sequence. The feature hooks should be before, they should register pre-processors and events on the rows and columns, and those should be called only once on mount, after `useGridRows` or `useGridColumns` are called.

To make it work, it seems that you just have to remove the `forceUpdate` in your new `updateCaches` method and to only run it in the `useEffect` below.
The `forceUpdate` is not needed when the `updateCaches` is called from the `rowsSet` event because `useGridRows` take care to trigger a new re-render after the event sequence.

```ts
  const updateCaches = React.useCallback(() => {
    setGridState((state) => {
      return {
        ...state,
        detailPanel: {
          ...state.detailPanel,
          ...cacheContentAndHeight(apiRef, state, getDetailPanelContent, getDetailPanelHeight),
        },
      };
    });
  }, [apiRef, forceUpdate, getDetailPanelContent, getDetailPanelHeight, setGridState]);

  useGridApiEventHandler(apiRef, GridEvents.rowsSet, updateCaches);

  React.useEffect(() => {
    updateCaches();
      forceUpdate();
  }, [updateCaches, forceUpdate]);
```

I missed it on the dedicated PR but you have the exact same thing for `useGridColumnPinning`.
All the tests pass if I move it next to `useGridTreeData` and that allows to have 1 `createColumnsState` instead of 2.
Each feature hook after `useGridColumns` causes a full pre-processing sequence which can be heavy because it causes the recomputation of the flex values.
(this can be fixed in another PR of course)

Reply:
I reordered the hook as you asked. However it was not a simple task. Since I updated to also compute the cache synchronously, the `useGridDetailPanel` hook couldn't be before `useGridRows`. To solve that, I moved the caching logic to another hook that runs after the rows are ready.

Reply:
Yeah as soon as we starting using `getRowParams` it limit a lot what we can do...
We could probably have a workaround based on events to trigger the 1st one when the rows changes.
But I don't thinks it's worth the overhead right now.

Closing this one :+1: